kernel/src/expressions/scalars.rs

```diff
@@ -3,10 +3,86 @@ use std::fmt::{Display, Formatter};
 
 use chrono::{DateTime, NaiveDate, NaiveDateTime, TimeZone, Utc};
 
-use crate::schema::{ArrayType, DataType, PrimitiveType, StructField};
+use crate::schema::{ArrayType, DataType, DecimalType, PrimitiveType, StructField};
 use crate::utils::require;
 use crate::{DeltaResult, Error};
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct DecimalValue {
+    bits: i128,
+    ty: DecimalType,
+}
+
+impl DecimalValue {
+    pub fn try_new(bits: impl Into<i128>, ty: DecimalType) -> DeltaResult<Self> {
+        let bits = bits.into();
+        require!(
+            ty.precision() >= get_decimal_precision(bits),
+            Error::invalid_decimal(format!(
+                "Decimal value {} exceeds precision {}",
+                bits,
+                ty.precision()
+            ))
+        );
+        Ok(Self { bits, ty })
+    }
+
+    pub fn bits(&self) -> i128 {
+        self.bits
+    }
+
+    pub fn ty(&self) -> &DecimalType {
+        &self.ty
+    }
+
+    pub fn precision(&self) -> u8 {
+        self.ty.precision()
+    }
+
+    pub fn scale(&self) -> u8 {
+        self.ty.scale()
+    }
+}
+
+/// Computes the decimal precision of a 128-bit number. The largest possible magnitude is i128::MIN
+/// = -2**127 with 39 decimal digits.
+fn get_decimal_precision(value: i128) -> u8 {
+    let mut precision = 0;
+
+    let mut value = value.unsigned_abs();
+    if value >= 1000_0000_0000_0000_0000_0000 {
+        value /= 1_0000_0000_0000_0000_0000_0000;
+        precision += 24;
+        // at most 39-24 = 15 digits remain
+    } else if value >= 1000_0000_0000 {
+        value /= 1_0000_0000_0000;
+        precision += 12;
+        // at most 24-12 = 12 digits remain
+    }
+
+    if value >= 1000_0000 {
+        value /= 1_0000_0000;
+        precision += 8;
+        // at most 7 decimal digits remain
+    }
+    if value >= 1000 {
+        value /= 1_0000;
+        precision += 4;
+        // at most 3 decimal digits remain
+    }
+    if value >= 10 {
+        value /= 1_00;
+        precision += 2;
+        // at most 1 decimal digit remains
+    }
+    if value >= 1 {
+        precision += 1;
+        // no digits left
+    }
+
+    precision
+}
```

Comment:
I _think_ we can just do:
```suggestion
fn get_decimal_precision(value: i128) -> u8 {
  value.checked_ilog10().map(|v| v + 1).unwrap_or(0) as u8
}
```

Reply:
Heh. I forgot the first rule: "don't optimize" :P